KUDO Zookeeper Platform Services Docs (#3530)
mesosphere/dcos-docs-site: pages/dkp/kommander/1.3/projects/platform-services/platform-services-catalog/kudo-zookeeper/index.md

---
layout: layout.pug
navigationTitle: KUDO Zookeeper
render: mustache
title: KUDO Zookeeper
menuWeight: 3
excerpt: Day 2 Operations of Apache Zookeeper
beta: false
---

<!-- markdownlint-disable MD018 -->
## KUDO Zookeeper

Kommander Catalog adds integration for [KUDO Zookeeper Operator](https://github.com/kudobuilder/operators/tree/master/repository/zookeeper), which simplifies day 2 operations of [Apache Zookeeper](https://zookeeper.apache.org/). 

#include /dkp/kommander/1.3/include/kudo-intro.tmpl

It is **strongly recommended** to view the [KUDO Zookeeper Documentation](https://github.com/kudobuilder/operators/tree/master/repository/zookeeper/docs/latest) which covers the KUDO Zookeeper Operator in-depth. This document covers the integration aspects of KUDO Zookeeper Operator with D2iQ Kommander.

### Kommander Catalog

KUDO Zookeeper is located in the Kommander Catalog.
#include /dkp/kommander/1.3/include/kommander-catalog-drilldown.tmpl

### Installation

> From the [Project Catalog](/dkp/kommander/1.3/projects/platform-services/) select the desired version of Zookeeper and click Deploy.

Reviewer: ```suggestion
From the [Project Catalog](/dkp/kommander/1.3/projects/platform-services/), select the desired version of Zookeeper and click Deploy.
```

> Here is an example of what should appear in the Kommander UI, the dialog is populated with sensible defaults:

Reviewer: Do we want this to say sensible?

![Zookeeper Service Install Configuration](/dkp/kommander/1.3/img/platform-services-zookeeper-config-dialog.png)

> - The **ID** field above is referred to as the Zookeeper `instance` by KUDO.

Reviewer 2: Not sure if this is really an in-scope comment on this doc, but when I last played with this stuff, I renamed the zookeeper instance on install but then I was super confused as to what the Zookeeper URI was.   You need to know this, of course, for other services. 

It would be nice if this were documented better in the external access section below, as there is no mention of the service names there, and a reference to the external access section added here to tie the two together?  (Something like `The instance is part of the the external name of the service; see the external access section for more details?`)

Author: @cneth I've now added docs to make this clearer for both KUDO ZK and also Kafka in the associated PR in #3532

Reviewer 2: Hi @kaiwalyajoshi - thanks, that's a big improvement!

- KUDO will then install Zookeeper in the Project namespace created using Kommander.
- The deployment progression can be viewed by looking at the `deploy` plan.

### KUDO Zookeeper Service Status

The Operator Plan Status provides an overview of the service.

At this point it is useful to have an understanding of [KUDO Operator Plans](https://kudo.dev/docs/what-is-kudo.html#operator-plans) and the [KUDO CLI](https://kudo.dev/docs/cli/installation.html) should be installed.

- Get the instances running in the project namespace.

#include /dkp/kommander/1.3/include/kudo-zookeeper-instance.tmpl

- Get the Plan status from the KUDO Zookeeper Operator.

#include /dkp/kommander/1.3/include/kudo-zookeeper-plans.tmpl

 - A `deploy` plan status of `COMPLETE` indicates that KUDO Zookeeper has deployed successfully and is healthy.

If any issues are encountered during the above, refer to the [Troubleshooting](#Troubleshooting) section.

### Available Parameters

The complete list of KUDO Zookeeper Parameters can be found under [detailed parameter descriptions](https://github.com/kudobuilder/operators/blob/master/repository/zookeeper/operator/params.yaml).

> 

Reviewer: don't need this extra break.
```suggestion
```

The current parameters set can be retrieved using the kubectl command with the two additional tools:
- [jq](https://stedolan.github.io/jq)
- [yq](https://mikefarah.gitbook.io/yq)

> To retrieve the current parameters, issue the following command in the terminal with appropriate `INSTANCE` value set:

Reviewer: ```suggestion
To retrieve the current parameters, use the following command in the terminal with appropriate `INSTANCE` value set:
```

```
INSTANCE=zookeeper;
kubectl -n test-project-zc6tc get instances -o json | jq ".items[] | select(.metadata.name == \"$INSTANCE\") | .spec.parameters" | yq -e --yaml-output '.' > zookeeper-params.yml
```

The above command generates a file called `zookeeper-params.yml` with the current values of all the parameters in effect for the `zookeeper` instance.

### Updating Parameters
Parameters can be updated using arguments to the KUDO CLI.

**Example**: Increasing Zookeeper node counts
- Increase the number of nodes using the KUDO CLI:
```
kubectl kudo update --instance zookeeper -p NODE_COUNT=5 -n test-project-zc6tc
```
- Monitor the KUDO Zookeeper deployment plan:
```
kubectl kudo plan status --instance zookeeper -n test-project-zc6tc
```
- Wait for the deployment plan to have a status of `COMPLETE`

> When the deployment plan is `COMPLETE` there should be 5 nodes as seen by the number of pods running:

Reviewer: ```suggestion
When the deployment plan is `COMPLETE`, there should be 5 nodes as seen by the number of pods running:
```

```
$ kubectl get pods -n test-project-zc6tc
NAME                    READY   STATUS    RESTARTS   AGE
zookeeper-zookeeper-0   1/1     Running   0          37s
zookeeper-zookeeper-1   1/1     Running   0          2m47s
zookeeper-zookeeper-2   1/1     Running   0          3m48s
zookeeper-zookeeper-3   1/1     Running   0          5m15s
zookeeper-zookeeper-4   1/1     Running   0          5m15s
```

**Example**: Updating multiple parameters:

To update multiple parameters at once, it is recommended to submit the updated parameters using the KUDO CLI.

See [Available Parameters](#available-parameters) to get the full list of current parameters as a file.

Apply the desired updates in `zookeeper-params.yml` using the KUDO CLI:
```
kubectl kudo update -n test-project-zc6tc --instance=zookeeper -P zookeeper-params.yml 
```
Wait for the deployment plan to `COMPLETE` as shown in the node counts example.

### Limitations

> The following parameters are viewed as immutable by the operator after an instance has been deployed, any modifications to them will trigger the `not-allowed` plan:

Reviewer 2: ```suggestion
The following parameters are treated as immutable by the operator after an instance has been deployed; any modifications to them will trigger the `not-allowed` plan:
```

- `DISK_SIZE`
- `STORAGE_CLASS`

> While modifying the KUDO Zookeeper parameters for `DISK_SIZE` are not permitted, users can resize the associated [PVC](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/).

Reviewer 2: ```suggestion
While modifying the KUDO Zookeeper parameter for `DISK_SIZE` us not permitted, users can resize the associated [PVC](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/).
```

Switching between persistent & ephemeral storage classes is not supported for `STORAGE_TYPE`.

### External Access

KUDO Zookeeper creates two Kubernetes Services:
- [Client Service (CS)](https://github.com/kudobuilder/operators/blob/master/repository/zookeeper/operator/templates/services.yaml#L20-L34) - For connecting clients to Zookeeper.

> - [High Availability Services (HS)](https://github.com/kudobuilder/operators/blob/master/repository/zookeeper/operator/templates/services.yaml#L1-L18) - For connecting additional Zookeeper servers to the quorum in [replicated mode](https://zookeeper.apache.org/doc/current/zookeeperStarted.html).

Reviewer 2: I think it would be good to elaborate a bit here beyond just pointing at YAML, first explaining what the difference is between the 'CS' and the 'HS' services and then explaining precisely how the names translate to endpoints.   The Kudo Zookeeper docs do not seem to discuss this at all...

The example below mentions you use the 'CS' with the zkCLI but there is no mention of what the HS service is, and the YAML does not explain it either.  From experience, I believe you use the HS for the Zookeeper URI access string for connecting with Kafka/Cassandra, etc, and the URI looks something like this:

`<instance>-zookeeper-[0-n].<instance>-hs:2181` (??)

But we should document this here to make it clear and to avoid frustration as most people will probably rename these things from the default.

Author: @cneth well, there's two Services:
- One for the clients to talk to the ZK Quorum over port 2181.
- One for ZK servers to talk to each other over two other ports: 2888 for servers to listen to requests from each other, 3888 where leader election happens.

There's a mismatch between the docs and code on the KUDO Site:
`<instance>-zookeeper-[0-n].<instance>-hs:2181` doesn't line up with https://github.com/kudobuilder/operators/blob/master/repository/zookeeper/operator/templates/services.yaml#L20-L34

As it should be `<instance>-zookeeper-[0-n].<instance>-cs:2181`

Overall I understand where the confusion is coming from and I'll attempt to clarify this. Stay tuned.

JLine support is enabled
2021-04-15 01:12:32,113 [myid:127.0.0.1:2181] - INFO  [main-SendThread(127.0.0.1:2181):ClientCnxn$SendThread@1438] - Session establishment complete on server localhost/127.0.0.1:2181, session id = 0x200021ea3ed0003, negotiated timeout = 30000

WATCHER::

WatchedEvent state:SyncConnected type:None path:null
[zk: 127.0.0.1:2181(CONNECTED) 0]
[zk: 127.0.0.1:2181(CONNECTED) 0] ls /
[zookeeper]
```

### Troubleshooting
KUDO provides the ability to collect logs and other [diagnostics data](https://kudo.dev/docs/cli/examples.html#collecting-diagnostic-data) for debugging and for bug-reports.
```
kubectl kudo diagnostics collect --instance zookeeper -n test-project-zc6tc
```

The diagnostics data contains the following:

KUDO Environment
- Installed Manager and its logs.
- Service account and services.

Data for the specified Operator
- The Operator, OperatorVersion and Instance resources.
- Deployed resources from the operator.
- Logs from deployed pods

> To monitor all the events occurring in the namespace, its helpful to look at event log:

Reviewer: ```suggestion
To monitor all the events occurring in the namespace, it's helpful to look at event log:
```
